homeassistant/components/xiaomi_miio/switch.py

```diff
@@ -250,12 +342,156 @@ async def async_service_handler(service):
+                )
+
+        for feature in SWITCH_TYPES:
+            switch = SWITCH_TYPES[feature]
+            if feature & device_features and feature in SWITCH_TYPES:
+                entities.append(
+                    XiaomiGenericCoordinatedSwitch(
+                        f"{config_entry.title} {switch.name}",
+                        device,
+                        config_entry,
+                        f"{switch.short_name}_{config_entry.unique_id}",
+                        switch,
+                        coordinator,
+                    )
+                )
+
+    async_add_entities(entities)
+
+
+class XiaomiGenericCoordinatedSwitch(XiaomiCoordinatedMiioEntity, SwitchEntity):
+    """Representation of a Xiaomi Plug Generic."""
+
+    def __init__(self, name, device, entry, unique_id, switch, coordinator):
+        """Initialize the plug switch."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+
+        self._attr_icon = switch.icon
+        self._controller = switch
+        self._state = self._extract_value_from_attribute(
+            self.coordinator.data, self._controller.short_name
+        )
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        # On state change the device doesn't provide the new state immediately.
+        self._state = self._extract_value_from_attribute(
+            self.coordinator.data, self._controller.short_name
+        )
+        self.async_write_ha_state()
+
+    @property
+    def available(self):
+        """Return true when state is known."""
+        if (
+            super().available
+            and not self.coordinator.data.is_on
+            and not self._controller.available_with_device_off
+        ):
+            return False
+        return super().available
+
+    @property
+    def is_on(self):
+        """Return the current switch state."""
+        return self._state
+
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        value = getattr(state, attribute)
+        if isinstance(value, Enum):
+            return value.value
+
+        return value
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn on an option of the miio device."""
+        method = getattr(self, SERVICE_TO_METHOD[self._controller.service]["method_on"])
+        if await method():
+            # Write state back to avoid switch flips with a slow response
+            self._state = True
+            self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn on an option of the miio device."""
+        method = getattr(
+            self, SERVICE_TO_METHOD[self._controller.service]["method_off"]
+        )
+        if await method():
+            # Write state back to avoid switch flips with a slow response
+            self._state = False
+            self.async_write_ha_state()
+
+    async def async_set_buzzer_on(self) -> bool:
```

Comment:
I'm not following here. Why are we registering these custom services that all seem to have a binary mode? Why not just create normal switch entities for these features if all that is needed is to turn a feature on or off?

Reply:
Nevermind, I see now that we're not registering custom services for these handlers.